src/Keboola/StorageApi/S3Uploader.php

```diff
@@ -0,0 +1,217 @@
+    }
+
+    /**
+     * @param $bucket
+     * @param $key
+     * @param $acl
+     * @param $slices
+     * @param null $encryption
+     * @throws ClientException
+     */
+    public function uploadSlicedFile($bucket, $key, $acl, $slices, $encryption = null)
+    {
+        $preparedSlices = [];
+        foreach ($slices as $filePath) {
+            $preparedSlices[$filePath] = $key . basename($filePath);
+        }
+        $chunker = new Chunker($this->transferOptions->getChunkSize());
+        $chunks = $chunker->makeChunks($preparedSlices);
+        foreach ($chunks as $chunk) {
+            $this->upload($bucket, $acl, $chunk, null, $encryption);
+        }
+    }
+
+    /**
+     * @param $bucket
+     * @param $acl
+     * @param $key
+     * @param $filePath
+     * @param null $name
+     * @param null $encryption
+     * @throws ClientException
+     */
+    protected function putFile($bucket, $key, $acl, $filePath, $name = null, $encryption = null)
+    {
+        $fh = @fopen($filePath, 'r');
+        if ($fh === false) {
+            throw new ClientException("Error on file upload to S3: " . $filePath, null, null, 'fileNotReadable');
+        }
+        $options = array(
+            'Bucket' => $bucket,
+            'Key' => $key,
+            'ACL' => $acl,
+            'Body' => $fh,
+            'ContentDisposition' => sprintf('attachment; filename=%s;', $name ? $name : basename($filePath)),
+        );
+
+        if ($encryption) {
+            $options ['ServerSideEncryption'] = $encryption;
+        }
+        $this->s3Client->putObject($options);
+        if (is_resource($fh)) {
+            fclose($fh);
+        }
+    }
+
+    /**
+     * @param $bucket
+     * @param $acl
+     * @param array $files
+     * @param null $name
+     * @param null $encryption
+     * @throws ClientException
+     */
+    protected function upload($bucket, $acl, $files, $name = null, $encryption = null)
+    {
+        $promises = [];
+        foreach ($files as $filePath => $key) {
+            /*
+             * Cannot upload empty files using multipart: https://github.com/aws/aws-sdk-php/issues/1429
+             * Upload them directly immediately and continue to next part in the chunk.
+             */
+            if (filesize($filePath) === 0) {
+                $this->putFile($bucket, $key, $acl, $filePath, $name, $encryption);
+                continue;
+            }
+            $uploader = $this->multipartUploaderFactory(
+                $filePath,
+                $bucket,
+                $key,
+                $acl,
+                count($files) > 1 ? $this->transferOptions->getMultiFileConcurrency() : $this->transferOptions->getSingleFileConcurrency(),
+                $encryption ? $encryption : null,
+                $name ? $name : basename($filePath)
+            );
+            $promises[$filePath] = $uploader->promise();
+        }
+
+        $retries = 0;
+        do {
```

Comment:
s tim finished je to tedka takova zvlastni konstrukce, neni jednodussi proste while (!$finished) { ... ?

Reply:
Jo, udělám, slepě jsem to vykopíroval z https://docs.aws.amazon.com/sdk-for-php/v3/developer-guide/s3-multipart-upload.html#recovering-from-errors

Reply:
A nespadá tohle do kategorie "líp se mi to čte"? :-)

Reply:
a když z toho `$finished` úplně vymlátím a z cyklu odejdu `break`em, tak to má o jedno odsazení míň.

Reply:
trosku spada :), ale vidis, ze se to objektivne zjednodusilo :)